[SPARK-12340][SQL] Fix overstep the bounds of Int in SparkPlan.executeTake

QiangCai opened the pull request:
If partsScanned = 1, numPartsToTry = Int.MaxValue and totalParts = 200, the result of math.min(partsScanned + numPartsToTry, totalParts) will be Int.MinValue (-2147483648).

Modifies partsScanned to partsScanned.toLong and  chang result of math.min to Int.

https://issues.apache.org/jira/browse/SPARK-12340

QiangCai committed:
Fix overstep the bounds of Int in SparkPlan.executeTake

Modifies partsScanned to partsScanned.toLong and  chang result of math.min to Int.

3ourroom commented:
NAVER - http://www.naver.com/
--------------------------------------------

3ourroom@naver.com 님께 보내신 메일 <[spark] [SPARK-12340][SQL] Fix overstep the bounds of Int in SparkPlan.executeTake (#10310)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

AmplabJenkins commented:
Can one of the admins verify this patch?

3ourroom commented:
NAVER - http://www.naver.com/
--------------------------------------------

3ourroom@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-12340][SQL] Fix overstep the bounds of Int in SparkPlan.executeTake (#10310)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -206,7 +206,7 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
 
       val left = n - buf.size
-      val p = partsScanned until math.min(partsScanned + numPartsToTry, totalParts)
+      val p = partsScanned until math.min(partsScanned.toLong + numPartsToTry, totalParts).toInt
```
Sounds good to me.

SparkQA commented:
**[Test build #2217 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2217/consoleFull)** for PR 10310 at commit [`37acd54`](https://github.com/apache/spark/commit/37acd541e0447a46a3b7718a688c4312e86d13e6).

SparkQA commented:
**[Test build #2217 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2217/consoleFull)** for PR 10310 at commit [`37acd54`](https://github.com/apache/spark/commit/37acd541e0447a46a3b7718a688c4312e86d13e6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class WrapOption(child: Expression) extends UnaryExpression `\n

QiangCai committed:
Fix overstep the bounds of Int in SparkPlan.executeTake

partsScanned maybe overstep the bounds of Int 

If partsScanned  = 1 and numPartsToTry  = Int.MaxValue, partsScanned += numPartsToTry  well be Int.MinValue. So we chang to use the size of  seq p.

QiangCai commented:
Tthe number of  partitions which has scanned is  the size of seq p, not numPartsToTry.

srowen commented:
Yes I think that also makes sense, in the case where we hit the limit of `totalParts`.

SparkQA commented:
**[Test build #2220 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2220/consoleFull)** for PR 10310 at commit [`56f151a`](https://github.com/apache/spark/commit/56f151a6193a8eee6bafa6fb224648271254b34f).

SparkQA commented:
**[Test build #2220 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2220/consoleFull)** for PR 10310 at commit [`56f151a`](https://github.com/apache/spark/commit/56f151a6193a8eee6bafa6fb224648271254b34f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

rxin commented:
Can you also change take in RDD?

QiangCai commented:
Yes. I will change all.

QiangCai commented:
Yes. I will check all at first.

srowen commented:
@QiangCai I think that technically `var partsScanned = 0` in `take` is a problem since it's incremented by `numPartsToTry` and could overflow causing `partsScanned < totalParts` to always be true. This is a very rare case where `numPartsToTry` is large and `totalParts` is close to the max integer, and `n` is very large.

QiangCai committed:
avoid Int overflow

QiangCai commented:
@srowen I have modified all the code, and try to keep them to be the same.

At first, the var numPartsToTry and partsScanned  have been set to Long.
Secondly,  because  var p should be Seq[Int], so i have added to.Int  in the until statement

srowen commented:
@QiangCai thanks that looks good, but this needs a rebase now.

QiangCai commented:
@srowen I maybe have made a mistake in git bash. Can I new another pull request to resolved this problem.

srowen commented:
I think you just need to rebase from master and force-push the result, but do what you need to.

QiangCai commented:
@srowen I have new another pull request. This pull request will be closed.
another pull request https://github.com/apache/spark/pull/10487
